### PR TITLE
refactor: recipient field

### LINKED ISSUE #3366: Refactor address vs. BNS name entry for recipient field in send form
This issue builds on https://github.com/hirosystems/stacks-wallet-web/issues/3364. Reusing the same field for bns names and recipients has proven difficult and error prone.

CleanShot 2023-03-06 at 10 00 27@2x

While we can fix the implementation, we've spent too much time on this, and the consequences of it going wrong are severe enough that, we should also reconsider the design of this component, provided that will simplify the implementation without a huge UX sacrifice.

@mica000 would be great if you could explore a few alternative UI for this for us to discuss & consider.

<details>
<summary>Some ideas</summary>

1) Having a toggle/tab interface where it's either recipient or bns name (not both). UI could prompt users entering a BNS name and even auto-switch to the bns tab if a name is detected.

2) A separate "Look up BNS name" flow, opening a modal with a search box, and the ability to look up a name. When found, the recipients actual address is entered to the recipient field.

</details>

### PR DESCRIPTION
**Pull request by fbwoolf**
> Try out this version of the Hiro Wallet - download [extension builds](https://github.com/hirosystems/wallet/actions/runs/4505475214).

Opening this as a draft PR to get feedback on the implementation here in the STX form. Once the implementation is agreeed on, I will implement in the BTC form and Stacks Sip10.

Screen Shot 2023-03-20 at 8 56 18 AM

Screen Shot 2023-03-19 at 7 20 21 PM

### CODE REVIEW

**Comment by kyranjamie** [comment]
CleanShot 2023-03-20 at 17 54 33@2x

File: src/app/pages/send/send-container.tsx [hunk 1]
```diff
@@ -24,6 +24,7 @@ export function SendContainer() {
         maxHeight="90vh"
         maxWidth={['100%', CENTERED_FULL_PAGE_MAX_WIDTH]}
         minWidth={['100%', CENTERED_FULL_PAGE_MAX_WIDTH]}
+        zIndex={9999}
```
Comment (kyranjamie) [inline src/app/pages/send/send-container.tsx]: What elements are above this? Is there a better way to fix?

File: src/app/pages/send/send-crypto-asset-form/components/field-error.tsx [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        <Flex height={openHeight + 'px'}>
+          <ErrorLabel data-testid={SendCryptoAssetSelectors.FormFieldInputErrorLabel} fontSize={1}>
+            {meta.error}
+          </ErrorLabel>
+        </Flex>
+      </AnimateHeight>
+    </Flex>
+  ) : (
+    <Box height={closedHeight + 'px'}></Box>
+  );
```
Comment (kyranjamie) [inline src/app/pages/send/send-crypto-asset-form/components/field-error.tsx]: early return easier to read

File: src/app/pages/send/send-crypto-asset-form/components/footer.tsx [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+  const isPopup = isPopupMode();
+
+  return (
+    <Box
+      bg={color('bg')}
+      borderTop="1px solid #DCDDE2"
+      bottom="0px"
+      height={['132px', '116px']}
+      position={isPopup ? 'absolute' : 'unset'}
```
Comment (kyranjamie) [inline src/app/pages/send/send-crypto-asset-form/components/footer.tsx]: Could also use `whenPageMode` util

File: src/app/pages/send/send-crypto-asset-form/components/recipient-select/recipient-select-item.tsx [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+      _hover={{ bg: isVisible ? color('bg-alt') : 'none', borderRadius: '8px' }}
+      alignItems="center"
+      bg={color('bg')}
+      height="32px"
+      isInline
+      mb="0px !important"
+      minWidth="110px"
+      onClick={() => onSelectItem(index)}
+      pl={isVisible ? 'tight' : 'unset'}
+    >
```
Comment (kyranjamie) [inline src/app/pages/send/send-crypto-asset-form/components/recipient-select/recipient-select-item.tsx]: ⚠️ no a clickable element

File: src/app/pages/send/send-crypto-asset-form/components/recipient-select/recipient-select.layout.tsx [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  selectedItem: number;
+}
+export function RecipientSelectLayout(props: RecipientSelectLayoutProps) {
+  const { children, isVisible, onSetIsSelectVisible, selectedItem } = props;
+  const ref = useRef<HTMLDivElement | null>(null);
+
+  useOnClickOutside(ref, () => onSetIsSelectVisible(false));
+
+  return (
+    <Stack _hover={{ cursor: 'pointer' }} maxHeight="base-loose">
```
Comment (kyranjamie) [inline src/app/pages/send/send-crypto-asset-form/components/recipient-select/recipient-select.layout.tsx]: Is this the correct element to use? Why have a cursor pointer on an element without an action? Styles likely belong elsewhere

File: src/app/pages/send/send-crypto-asset-form/components/recipient-select/recipient-select.layout.tsx [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+            ref={ref}
+            style={styles}
+            top="40px"
+            zIndex={9999}
+          >
+            {children}
+          </Stack>
+        )}
+      </Fade>
+    </Stack>
```
Comment (kyranjamie) [inline src/app/pages/send/send-crypto-asset-form/components/recipient-select/recipient-select.layout.tsx]: I know you're reluctant to use, and not suggesting to change rn, but a dropdown component would be easier implementation, and have better ux.

Consider UX behaviours: advanced mouse controls, keyboard controls (esc to close, arrow keys to select) etc.

File: src/app/pages/send/send-crypto-asset-form/components/recipient-select/recipient-select.tsx [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+        onSelectItem={onSelectItem}
+      />
+      <RecipientSelectItem
+        index={RecipientFieldType.BnsName}
+        isVisible={isVisible}
+        onSelectItem={onSelectItem}
+      />
+    </RecipientSelectLayout>
+  );
+}
```
Comment (kyranjamie) [inline src/app/pages/send/send-crypto-asset-form/components/recipient-select/recipient-select.tsx:36]: Good composition 👍🏼

File: src/app/pages/send/send-crypto-asset-form/family/stacks/components/stacks-recipient-field/components/recipient-bns-address-displayer.tsx [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+        <Stack>
+          <Box
+            _hover={{ cursor: 'pointer' }}
+            as={FiCopy}
+            color={color('text-caption')}
+            data-testid={SendCryptoAssetSelectors.RecipientBnsAddressCopyToClipboard}
+            onClick={copyToClipboard}
+            size="16px"
+          />
+        </Stack>
```
Comment (kyranjamie) [inline src/app/pages/send/send-crypto-asset-form/family/stacks/components/stacks-recipient-field/components/recipient-bns-address-displayer.tsx]: What does the stack do here?

File: src/app/pages/send/send-crypto-asset-form/family/stacks/components/stacks-recipient-field/components/recipient-bns-address-displayer.tsx [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+      <Tooltip hideOnClick={false} label={hasCopied ? 'Copied!' : 'Copy address'} placement="right">
+        <Stack>
+          <Box
+            _hover={{ cursor: 'pointer' }}
+            as={FiCopy}
+            color={color('text-caption')}
+            data-testid={SendCryptoAssetSelectors.RecipientBnsAddressCopyToClipboard}
+            onClick={copyToClipboard}
+            size="16px"
+          />
```
Comment (kyranjamie) [inline src/app/pages/send/send-crypto-asset-form/family/stacks/components/stacks-recipient-field/components/recipient-bns-address-displayer.tsx]: ⚠️ not a clickable element

File: src/app/pages/send/send-crypto-asset-form/family/stacks/components/stacks-recipient-field/stacks-recipient-field.tsx [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+      topInputOverlay={topInputOverlay}
+    />
+  ) : (
+    <RecipientFieldBnsName
+      isSelectVisible={isSelectVisible}
+      onClickLabelAction={onClickLabelAction}
+      selectedRecipientField={selectedRecipientField}
+      topInputOverlay={topInputOverlay}
+    />
+  );
```
Comment (kyranjamie) [inline src/app/pages/send/send-crypto-asset-form/family/stacks/components/stacks-recipient-field/stacks-recipient-field.tsx]: switch statement?

**Review by kyranjamie — COMMENTED** [review]
Great work Fara, written down some thoughts. 

Noticed a few places where there are click handers with the wrong semantics. React makes way too easy to completely forget we're writing HTML, would love to add validation to catch these. Caring about accessibility isn't just to help blind people using screen readers. It helps helpers power users, and generally improves the user experience of the web  ✨

File: src/app/pages/send/send-container.tsx (hunk 1 above)
Comment (fbwoolf) [reply src/app/pages/send/send-container.tsx]: I believe the footer needed a Zindex which caused this to need to be about that.

File: src/app/pages/send/send-crypto-asset-form/components/footer.tsx (hunk 3 above)
Comment (fbwoolf) [reply src/app/pages/send/send-crypto-asset-form/components/footer.tsx]: ~~When I used `whenPageMode` it seemed a bit like overkill when just one prop was different?~~

File: src/app/pages/send/send-crypto-asset-form/components/recipient-select/recipient-select-item.tsx (hunk 4 above)
Comment (fbwoolf) [reply src/app/pages/send/send-crypto-asset-form/components/recipient-select/recipient-select-item.tsx]: I missed this, and just something I've become aware of, so I'll fix.

File: src/app/pages/send/send-crypto-asset-form/components/recipient-select/recipient-select.layout.tsx (hunk 5 above)
Comment (fbwoolf) [reply src/app/pages/send/send-crypto-asset-form/components/recipient-select/recipient-select.layout.tsx]: I'll double check the select code. Most of is was copied over from the fees selector.

Comment (fbwoolf) [reply src/app/pages/send/send-crypto-asset-form/components/recipient-select/recipient-select.layout.tsx]: Maybe I should at least make the select code shareable ...but tbh, I hate spending time on UI components too much right now.

File: src/app/pages/send/send-crypto-asset-form/family/stacks/components/stacks-recipient-field/components/recipient-bns-address-displayer.tsx (hunk 8 above)
Comment (fbwoolf) [reply src/app/pages/send/send-crypto-asset-form/family/stacks/components/stacks-recipient-field/components/recipient-bns-address-displayer.tsx]: Tbh, I don't quite get it either. I removed it and the tooltip didn't work. Let me try again. I've seen this in several places.

**Comment by markmhendrickson** [comment]
This is looking good! 👏

File: src/app/pages/send/send-container.tsx (hunk 1 above)
Comment (fbwoolf) [reply src/app/pages/send/send-container.tsx]: I think this might have been just left over from me testing things out. It appears I can remove it.

File: src/app/pages/send/send-crypto-asset-form/components/footer.tsx (hunk 3 above)
Comment (fbwoolf) [reply src/app/pages/send/send-crypto-asset-form/components/footer.tsx]: I see what you mean here now, nvm.

File: src/app/pages/send/send-crypto-asset-form/family/stacks/components/stacks-recipient-field/components/recipient-bns-address-displayer.tsx (hunk 8 above)
Comment (fbwoolf) [reply src/app/pages/send/send-crypto-asset-form/family/stacks/components/stacks-recipient-field/components/recipient-bns-address-displayer.tsx]: Was literally bc of `as={FiCopy}` here I think. Wrapped `FiCopy` in a div and made it clickable `as="button"` ...Stack is not needed. Will fix if I see other places.

**Comment by fbwoolf** [comment]
Going to move on here and implement with other forms, and then I will open the PR.

File: src/app/pages/send/send-crypto-asset-form/components/footer.tsx [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+import { formatMoney } from '@app/common/money/format-money';
+import { whenPageMode } from '@app/common/utils';
+import { SpaceBetween } from '@app/components/layout/space-between';
+import { Caption } from '@app/components/typography';
+import { HighFeeDrawer } from '@app/features/high-fee-drawer/high-fee-drawer';
+
+import { PreviewButton } from './preview-button';
+
+export function Footer(props: { balance: Money; url: string }) {
```
Comment (kyranjamie) [inline src/app/pages/send/send-crypto-asset-form/components/footer.tsx]: CryptoFormFooter?

**Comment by fbwoolf** [comment]
Opening this up for review.

File: src/app/pages/send/ordinal-inscription/components/form-fields.layout.tsx [hunk 12]
```diff
@@ -0,0 +1,12 @@
+import { Flex } from '@stacks/ui';
+
+interface FormFieldsProps {
+  children: React.ReactNode;
+}
+export function FormFieldsLayout({ children }: FormFieldsProps) {
```
Comment (kyranjamie) [inline src/app/pages/send/ordinal-inscription/components/form-fields.layout.tsx]: These names will get confusing, we might have 10x of these, should be more specific

File: src/app/pages/send/send-crypto-asset-form/components/form-footer.tsx [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+    >
+      <Box mt="loose" px="loose">
+        <PreviewButton />
+        <SpaceBetween>
+          <Caption>Balance</Caption>
+          <Caption>{formatMoney(balance)}</Caption>
+        </SpaceBetween>
+      </Box>
+      <HighFeeDrawer learnMoreUrl={url} />
+      <Outlet />
```
Comment (kyranjamie) [inline src/app/pages/send/send-crypto-asset-form/components/form-footer.tsx]: Why is the outlet in the footer?

File: src/app/pages/send/send-crypto-asset-form/components/recipient-select/components/recipient-select-item.tsx [hunk 14]
```diff
@@ -0,0 +1,37 @@
+import { FiChevronDown } from 'react-icons/fi';
+
+import { Box, Text, color } from '@stacks/ui';
+
+const labels = ['Address', 'BNS Name'];
+const testLabels = ['address', 'bns-name'];
+
+interface RecipientSelectItemProps {
+  index: number;
```
Comment (kyranjamie) [inline src/app/pages/send/send-crypto-asset-form/components/recipient-select/components/recipient-select-item.tsx:9]: Seems like an odd API choice. How am I supposed to know that `1` === `bns-name` without diving into the implementation? 

Why not pass in a typed string that looks up a keyed object?

File: src/app/common/validation/forms/address-validators.ts [hunk 15]
```diff
@@ -21,16 +21,17 @@ export function notCurrentAddressValidator(currentAddress: string) {
 export function btcAddressValidator() {
   return yup
     .string()
-    .defined('Enter a Bitcoin address')
+    .defined(FormErrorMessages.AddressRequired)
```
Comment (edu-stx) [inline src/app/common/validation/forms/address-validators.ts:24]: Checking whether we want [defined](https://github.com/jquense/yup#schemadefined-schema) or [required](https://github.com/jquense/yup#schemarequiredmessage-string--function-schema) here?

**Review by edu-stx — APPROVED** [review]
👍

**Review by alter-eggo — APPROVED** [review]
great work👍

File: src/app/common/validation/forms/address-validators.ts (hunk 15 above)
Comment (fbwoolf) [reply src/app/common/validation/forms/address-validators.ts:24]: Good question, I just changed the error message itself. Is `required` better in this case over `defined`?

**Review by kyranjamie — APPROVED** [review]
Sorry for slow review, had written comments and forgot to post.

LGTM, but we should avoid names like `FormFields`. This is very non-descriptive and unhelpful when we have > 1 form in the wallet.

File: src/app/pages/send/ordinal-inscription/components/form-fields.layout.tsx (hunk 12 above)
Comment (fbwoolf) [reply src/app/pages/send/ordinal-inscription/components/form-fields.layout.tsx]: Yep, agree, made this quick temporarily bc @alter-eggo is really going to take on refactoring the file structure of these forms. The ordinal code is located in a weird in that it used components in the send form. I changed the other layout component and broke this form by accident. I just didn't want to spend much time restructuring/renaming bc that is really a task assigned to @alter-eggo.

File: src/app/pages/send/send-crypto-asset-form/components/form-footer.tsx (hunk 13 above)
Comment (fbwoolf) [reply src/app/pages/send/send-crypto-asset-form/components/form-footer.tsx]: They all the send forms need it, so why not just do it once? I can move it back into each form ...I get your point,  maybe, that this could be moved higher and reused? Typically this preview button will require nested routes?

File: src/app/pages/send/send-crypto-asset-form/components/recipient-select/components/recipient-select-item.tsx (hunk 14 above)
Comment (fbwoolf) [reply src/app/pages/send/send-crypto-asset-form/components/recipient-select/components/recipient-select-item.tsx:9]: Yeah, again, copied this over from the fee select and wan't spending much time on refactoring considering we will swap it out soon for a new UI library.

**Comment by fbwoolf** [comment]
@markmhx @mica000 I'm going to merge this into dev so it doesn't get behind. Please review the build there for recipient field / send form changes.

File: src/app/pages/send/send-crypto-asset-form/components/recipient-select/components/recipient-select-item.tsx (hunk 14 above)
Comment (fbwoolf) [reply src/app/pages/send/send-crypto-asset-form/components/recipient-select/components/recipient-select-item.tsx:9]: I addressed your other comment, but going to come back to this one bc I'll spend time on the headless select component we discussed in UI mtg. I'll swap this out hopefully soon. 🤞
